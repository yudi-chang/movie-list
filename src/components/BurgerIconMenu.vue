<template>
  <div class="hamburger-wrapper" :class="{ 'active': active }" id="icon" @click="toggleMenu()">
    <div class="icon-1" id="a"></div>
    <div class="icon-2" id="b"></div>
    <div class="icon-3" id="c"></div>
    <div class="clear"></div>
  </div> 
</template>

<script setup lang="ts">
// model
const active = defineModel('active', { required: true, type: Boolean })

// methods
const toggleMenu = (): void => {
  active.value = !active.value
};
</script>

<style scoped lang="scss">
.hamburger-wrapper {
  position: relative;
  height: 60px;
  width: 60px;
  cursor: pointer;
  transition: all 0.2s ease-in-out;
  background: $secondary-color;

  .icon-1, .icon-2, .icon-3 {
      position: absolute;
      left: 25%;
      top: 50%;
      width: 32px;
      height: 3px;
      background-color: $primary-color;
      transition: all 400ms cubic-bezier(.84,.06,.52,1.8);
  }

  .icon-1 {
    transform: translateY(-8px);
    animation-delay: 100ms;
  }

  .icon-3 {
    transform: translateY(8px);
    animation-delay: 250ms;
  }

  &.active {    
    .icon-1 {
      transform: rotate(40deg);
    }

    .icon-3 {
      transform: rotate(-40deg);
    }

    .icon-2 {
      opacity: 0;
    }
  }
}

.clear {
  clear: both;
}
</style>