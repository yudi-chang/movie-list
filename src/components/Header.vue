<template>
  <header>
    <h1 class="pv-8 text-3xl fw-bold" style="cursor: pointer;"><RouterLink to="/">Film DB</RouterLink></h1>
    <nav :class="{ 'active' : isActive }">
      <ul>
        <li 
          v-for="link in links" 
          :key="link.name" 
          :class="{ active: route.path === link.to }"
          class="fw-bold ph-12 pv-8"
        >
          <span>
            <RouterLink :to="link.to">{{ link.name }}</RouterLink>
          </span>
        </li>
      </ul>
    </nav>
    <BurgerIconMenu v-model:active="isActive" class="burger-icon-menu"/>
  </header>
</template>


<script setup lang="ts">
import { ref, watch, inject } from 'vue';
import type { Ref } from 'vue';
import { RouterLink, useRoute } from 'vue-router'
import BurgerIconMenu from './BurgerIconMenu.vue';

const route = useRoute();

// type declaration
type Link = {
  name: string;
  to: string;
};

// reactive vars
const isActive: Ref<boolean> = ref(false);

// other vars
// no need to make this reactive
const links: Link[] = [
  { name: "Movies", to: "/" },
  { name: "Contact Us", to: "/contact-us" }
];

// watch 
watch(
  () => route.fullPath,
  () => {
    // force close mobile menu when changing page
    isActive.value = false;
  }
);
</script>


<style lang="scss" scoped>
header {
  background-color: $primary-color;
  display: flex;
  justify-content: space-between;
  align-items: end;
  padding: 0 20px;

  .burger-icon-menu {
    display: none;
  }

  * {
    color: $secondary-color;
  }

  ul {
    list-style: none;

    li {
      display: inline-block;
      vertical-align: bottom;
      margin: 0 5px;
      font-size: 20px;
      position: relative;
      
      span {
        &:after {
          content: '';
          position: absolute;
          left: 50%;
          bottom: 0;
          width: 0;
          height: 4px;
          transform: translateX(-50%);
          background-color: $secondary-color;
          border-top-right-radius: 5px;
          border-top-left-radius: 5px;
          transition: width 0.3s ease;
        }
      }

      &.active, &:hover {

        span {
          &:after {
            width: 100%;
            transition: width 0.3s ease;
          }
        }
      }
    }
  }
}

@media (max-width: $breakpoint-sm) {
  header {
    padding-right: 0;

    nav {
      background: $primary-color;
      position: fixed;
      top: 60px;
      right: 0;
      height: 100%;
      width: 0%;
      z-index: 10;
      opacity: 0;
      transition: all 600ms cubic-bezier(.62,.04,.3,1.56);
      transition-delay: 100ms;
      border-top: solid 1px $secondary-color;
      padding-top: 20px;
      padding-right: 10px;
      
      * {
        color: $secondary-color;
      }
      
      &.active {
        width: 100%;
        opacity: 1;
      }

      ul {
        margin: 0;

        li {
          list-style: none;
          font-size: 24px;
          color: #fff;
          line-height: 2.2;
          text-transform: uppercase;
          letter-spacing: 1.7px;
          display: block;
          width: auto;
          text-align: right;

          span {
            position: relative;

            &:after {
              bottom: -8px;
            }
          }
        }
      }
    }

    .burger-icon-menu {
      display: block;
    }
  }
}
</style>