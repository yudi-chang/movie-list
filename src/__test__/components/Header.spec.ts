import { describe, it, expect, beforeEach, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import Header from '@/components/Header.vue';
import BurgerIconMenu from '@/components/BurgerIconMenu.vue';
import router from '@/router';

describe('Header', () => {
  const createWrapper = (params: any) => {
    return mount(Header, params);
  };

  beforeEach(async () => {
    await router.push('/');
    await router.isReady();
  });

  it('should render header correctly', async () => {
    const wrapper = createWrapper({
      global: {
        plugins: [
          router,
          createTestingPinia({
            stubActions: false,
            createSpy: vi.fn,
          }),
        ],
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should updates active class of nagivation item based curent route', async () => {
    await router.push('/favorites');
    await router.isReady();

    const wrapper = mount(Header, {
      global: {
        plugins: [router, createTestingPinia({
          stubActions: false,
          createSpy: vi.fn,
        })]
      }
    });

    const activeLink = wrapper.find('[data-test="navigation-item"].active');
    expect(activeLink.exists()).toBe(true);
    expect(activeLink.text()).toContain('Favorites');
  });

  it('should toggles mobile menu when BurgerIconMenu is clicked', async () => {
    const wrapper = mount(Header, {
      global: {
        plugins: [router, createTestingPinia({
          stubActions: false,
          createSpy: vi.fn,
        })]
      }
    });

    const burgerMenu = wrapper.findComponent(BurgerIconMenu);
    expect(wrapper.find('nav').classes()).not.toContain('active');

    await burgerMenu.vm.$emit('update:active', true);
    expect(wrapper.find('nav').classes()).toContain('active');

    await router.push('/contact-us');
    expect(wrapper.find('nav').classes()).not.toContain('active');
  });

  
});
