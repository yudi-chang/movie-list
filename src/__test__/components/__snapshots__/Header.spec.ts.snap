// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should render header correctly 1`] = `
<header
  data-v-c970699f=""
>
  <h1
    class="pv-8 text-3xl fw-bold"
    data-v-c970699f=""
    style="cursor: pointer;"
  >
    <a
      aria-current="page"
      class="router-link-active router-link-exact-active"
      data-v-c970699f=""
      href="/"
    >
      Film DB
    </a>
  </h1>
  <nav
    class=""
    data-v-c970699f=""
  >
    <ul
      data-v-c970699f=""
    >
      
      <li
        class="active fw-bold ph-12 pv-8"
        data-test="navigation-item"
        data-v-c970699f=""
      >
        <span
          data-v-c970699f=""
        >
          <a
            aria-current="page"
            class="router-link-active router-link-exact-active"
            data-v-c970699f=""
            href="/"
          >
            Movies 
            <!--v-if-->
          </a>
        </span>
      </li>
      <li
        class="fw-bold ph-12 pv-8"
        data-test="navigation-item"
        data-v-c970699f=""
      >
        <span
          data-v-c970699f=""
        >
          <a
            class=""
            data-v-c970699f=""
            href="/favorites"
          >
            Favorites 
            <div
              class="counter-wrapper text-xs"
              data-test="counter-wrapper"
              data-v-c970699f=""
              data-v-f9cbd204=""
            >
              0
            </div>
          </a>
        </span>
      </li>
      <li
        class="fw-bold ph-12 pv-8"
        data-test="navigation-item"
        data-v-c970699f=""
      >
        <span
          data-v-c970699f=""
        >
          <a
            class=""
            data-v-c970699f=""
            href="/contact-us"
          >
            Contact Us 
            <!--v-if-->
          </a>
        </span>
      </li>
      
    </ul>
  </nav>
  <div
    class="hamburger-wrapper burger-icon-menu"
    data-test="hamburger-menu"
    data-v-c970699f=""
    data-v-d2ad046a=""
    id="icon"
  >
    <div
      class="icon-1"
      data-v-d2ad046a=""
      id="a"
    />
    <div
      class="icon-2"
      data-v-d2ad046a=""
      id="b"
    />
    <div
      class="icon-3"
      data-v-d2ad046a=""
      id="c"
    />
    <div
      class="clear"
      data-v-d2ad046a=""
    />
  </div>
</header>
`;
